component/loki/source/windowsevent/component_windows.go

```diff
@@ -0,0 +1,148 @@
+package windowsevent
+
+import (
+	"context"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/common/loki"
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/scrapeconfig"
+	"github.com/grafana/loki/clients/pkg/promtail/targets/windows"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "loki.source.windowsevent",
+		Args: Arguments{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+var (
+	_ component.Component = (*Component)(nil)
+)
+
+// Component implements the loki.source.windowsevent component.
+type Component struct {
+	opts component.Options
+
+	mut       sync.RWMutex
+	args      Arguments
+	target    *windows.Target
+	handler   chan api.Entry
+	receivers []loki.LogsReceiver
+}
+
+func (c *Component) Chan() chan<- api.Entry {
+	return c.handler
+}
+
+func (c *Component) Stop() {
+	// This is a noop.
+}
+
+// New creates a new loki.source.windowsevent component.
+func New(o component.Options, args Arguments) (*Component, error) {
+
+	c := &Component{
+		opts:      o,
+		receivers: args.ForwardTo,
+		handler:   make(chan api.Entry),
+		args:      args,
+	}
+
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+		if c.target != nil {
+			_ = c.target.Stop()
+		}
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler:
+			lokiEntry := loki.Entry{
+				Labels: entry.Labels,
+				Entry:  entry.Entry,
+			}
+			for _, receiver := range c.receivers {
+				receiver <- lokiEntry
+			}
+		}
+	}
+
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	// If no bookmark specified create one in the datapath.
+	if newArgs.BookmarkPath == "" {
+		newArgs.BookmarkPath = path.Join(c.opts.DataPath, "bookmark.xml")
+	}
+
+	// Create the bookmark file and parent folders if they don't exist.
+	_, err := os.Stat(newArgs.BookmarkPath)
+	if os.IsNotExist(err) {
+		err := os.MkdirAll(path.Dir(newArgs.BookmarkPath), 644)
+		if err != nil {
+			return err
+		}
+		_, err = os.Create(newArgs.BookmarkPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	winTarget, err := windows.New(c.opts.Logger, c, nil, convertConfig(newArgs))
```

Comment:
Does the Promtail code create any internal labels, similar to our discussion around `loki.source.syslog`?

Since we've merged #2697, we can have an optional Argument for this, and if it's not nil then pass it through `ComponentToPromRelabelConfigs` and add it as the argument here ^^

Reply:
Just checked and it does not create and labels prefixed with `__`, it adds channel (event log source) and the computer name as regular labels.